test/test_sparse_csr.py

```diff
@@ -2364,56 +2364,145 @@ def run_test(shape, nnz, index_type):
     @dtypes(*all_types_and_complex_and(torch.half, torch.bool, torch.bfloat16))
-    def test_transpose(self, device, dtype):
-
-        def run_test(shape, nnz, index_type):
-            # CSR
-            a = self.genSparseCSRTensor(shape, nnz, dtype=dtype, device=device, index_dtype=index_dtype)
-            self.assertEqual(a.layout, torch.sparse_csr)
-
-            # CSC
-            a_t = a.transpose(0, 1)
-            self.assertEqual(a_t.layout, torch.sparse_csc)
-
-            # CSR
-            a_v = a.transpose(0, 0)
-            self.assertEqual(a_v.layout, torch.sparse_csr)
-
-            # CSR again
-            a_t_t = a_t.transpose(0, 1)
-            self.assertEqual(a_t_t.layout, torch.sparse_csr)
-
-            # TODO: Do we want to extend view properties to members as well?
-            # These checks are based on is_view_of from test_view_ops.py
-            self.assertTrue(a_t._is_view())
-            self.assertTrue(a_v._is_view())
-            self.assertTrue(a_t_t._is_view())
-
-            self.assertTrue(a_t._base is a)
-            self.assertTrue(a_v._base is a)
-            self.assertTrue(a_t_t._base is a)
-
-            self.assertFalse(a_t is a)
-            self.assertFalse(a_v is a)
-            self.assertFalse(a_t_t is a)
-
-            self.assertEqual(a.to_dense().transpose(0, 1), a_t.to_dense())
-            self.assertEqual(a.to_dense(), a_v.to_dense())
-            self.assertEqual(a.to_dense(), a_t_t.to_dense())
-
-            with self.assertRaisesRegex(RuntimeError, "torch.transpose_: in-place transposition is not supported"):
-                a.transpose_(0, 0)
-
-            with self.assertRaisesRegex(RuntimeError, "torch.transpose_: in-place transposition is not supported"):
-                a.transpose_(0, 1)
-
+    @all_sparse_compressed_layouts()
+    def test_transpose(self, device, dtype, layout):
 
-        for shape, index_dtype in itertools.product(
-                [(10, 5), (10, 10)],
-                [torch.int32, torch.int64]):
-            run_test(shape, 0, index_dtype)
-            run_test(shape, max(shape), index_dtype)
-            run_test(shape, shape[0] * shape[1], index_dtype)
+        def to_dense(t):
+            if t.layout == torch.sparse_bsc:
+                # workaround untill bsc->dense conversions land
+                return t.to_sparse_csc().to_dense()
+            else:
+                return t.to_dense()
+
+        def _check_transpose_view(subject, transpose):
+            self.assertTrue(transpose.values()._is_view())
+            self.assertTrue(transpose._is_view())
+            self.assertTrue(transpose._base is subject)
+
+        def _check_layout_invariants(transpose):
+            compressed_indices_mth, plain_indices_mth = sparse_compressed_indices_methods[transpose.layout]
+            compressed_indices, plain_indices = compressed_indices_mth(transpose), plain_indices_mth(transpose)
+            # note: invariant check for bsr/bsc values is incorrect
```

Comment:
not sure how to interpret 😬 is the test conforming to the incorrect invariant or, like, compensating for it? should we file an issue (or have we? if so maybe a link)

Reply:
It is compensating for the fact that the `_validate_sparse_compressed_tensor_args` method is too strict. That method requires `values.is_contiguous()` but , [the blog post](https://pearu.github.io/bsr_tensor_invariants.html) specifies `values.is_contiguous()` or `values.transpose(-2 - N, -1 - N).is_contiguous()` Where `N` is `n_dense` (see invariant 3.7). 

This workaround is actually a bit careless though as the transpose check on the values does not take into account the dense dims and just assumes block dims are the last two which is only the case for non-hybrid inputs.

Reply:
I just checked and there is a TODO there indicating that the check needs to be relaxed for blocked layouts. 

https://github.com/pytorch/pytorch/blob/d9676d036b29eaede27e1b98a1d11f286ebf7b6c/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L100